docs/docs/icicle/introduction.md

```diff
@@ -0,0 +1,260 @@
+cargo build --release
+```
+
+This may take a couple of minutes since we are compiling both the CUDA and Rust code.
+
+To run the tests
+
+```sh
+cargo test
+```
+
+We also include some benchmarks
+
+```sh
+cargo bench
+```
+
+#### ICICLE Golang
+
+Golang is WIP in v1, coming soon. Please checkout a previous [release v0.1.0](https://github.com/ingonyama-zk/icicle/releases/tag/v0.1.0) for golang bindings.
+
+### Running ICICLE examples
+
+ICICLE examples can be found [here](https://github.com/ingonyama-zk/icicle-examples) these examples cover some simple use cases using C++, rust and golang.
+
+In each example directory, ZK-container files are located in a subdirectory `.devcontainer`.
+
+```sh
+msm/
+├── .devcontainer
+   ├── devcontainer.json
+   └── Dockerfile
+```
+
+Lets run one of our C++ examples, in this case the [MSM example](https://github.com/ingonyama-zk/icicle-examples/blob/main/c%2B%2B/msm/example.cu).
+
+Clone the repository
+
+```sh
+git clone https://github.com/ingonyama-zk/icicle-examples.git
+cd icicle-examples
+```
+
+Enter the test directory
+
+```sh
+cd c++/msm
+```
+
+Now lets build our docker file and run the test inside it. Make sure you have installed the [optional prerequisites](#optional-prerequisites).
+
+```sh
+docker build -t icicle-example-msm -f .devcontainer/Dockerfile .
+```
+
+Lets start and enter the container
+
+```sh
+docker run -it --rm --gpus all -v .:/icicle-example icicle-example-msm
+```
+
+to run the example
+
+```sh
+rm -rf build
+mkdir -p build
+cmake -S . -B build
+cmake --build build
+./build/example
+```
+
+You can now experiment with our other examples, perhaps try to run a rust or golang example next.
+
+## Writing new bindings for ICICLE
+
+Since ICICLE Core is written in CUDA / C++ its really simple to generate static libraries. These static libraries can be installed on any system and called by higher level languages such as Golang.
+
+static libraries can be loaded into memory once and used by multiple programs, reducing memory usage and potentially improving performance. They also allow you to separate functionality into distinct modules so your static library may need to compile only specific features that you want to use.
+
+Lets review the Golang bindings since its a pretty verbose example (compared to rust which hides it pretty well) of using static libraries. Golang has a library named `CGO` which can be used to link static libraries. Here's a basic example on how you can use cgo to link these libraries:
+
+```go
+/*
+#cgo LDFLAGS: -L/path/to/shared/libs -lbn254 -lbls12_381 -lbls12_377 -lbw6_671
+#include "icicle.h" // make sure you use the correct header file(s)
+*/
+import "C"
+
+func main() {
+  // Now you can call the C functions from the ICICLE libraries.
+  // Note that C function calls are prefixed with 'C.' in Go code.
+
+  out := (*C.BN254_projective_t)(unsafe.Pointer(p))
+  in := (*C.BN254_affine_t)(unsafe.Pointer(affine))
+
+  C.projective_from_affine_bn254(out, in)
+}
```

Comment:
I'm not sure how useful this example is on it's own.  Ideally we should have two high level walk-throughs.  

1.  Running Icicle Go Wrappers
2.  Directly running the C lib

My intuition for 2. is this will be an edge case and will need more background than just the provided code.

Reply:
Im adding this in another pr with golang wrapper docs